### PR TITLE
8291302: ARM32: nmethod entry barriers support

### PR DESCRIPTION
**Pull request by voitylov**
This PR implements nmethod entry barriers for ARM32. It has already been implemented for other ports and is related with JDK-8290025 "Remove the Sweeper".

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8291302](https://bugs.openjdk.org/browse/JDK-8291302): ARM32: nmethod entry barriers support

### Reviewers
 * [Erik Österlund](https://openjdk.org/census#eosterlund) (@fisk - **Reviewer**)
 * [Richard Reingruber](https://openjdk.org/census#rrich) (@reinrich - **Reviewer**) ⚠️ Review applies to [079a1a9b](https://git.openjdk.org/jdk/pull/11442/files/079a1a9b79a13220c8b629ebf327e02a80634c8e)
 * [Martin Doerr](https://openjdk.org/census#mdoerr) (@TheRealMDoerr - **Reviewer**)
 * [Andrew Haley](https://openjdk.org/census#aph) (@theRealAph - **Reviewer**) ⚠️ Review applies to [079a1a9b](https://git.openjdk.org/jdk/pull/11442/files/079a1a9b79a13220c8b629ebf327e02a80634c8e)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/11442/head:pull/11442` \
`$ git checkout pull/11442`

Update a local copy of the PR: \
`$ git checkout pull/11442` \
`$ git pull https://git.openjdk.org/jdk pull/11442/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 11442`

View PR using the GUI difftool: \
`$ git pr show -t 11442`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/11442.diff">https://git.openjdk.org/jdk/pull/11442.diff</a>

</details>

### CODE REVIEW

**Comment by voitylov** [comment]
Errors in pre-submit tests are clearly unrelated, GHA can't download the bundles for linux-x86:

Error: Unable to find an artifact with the name: bundles-linux-x86

**Review by fisk — CHANGES_REQUESTED** [review]
When you call the nmethod entry barrier slow path, floating point arguments are live, but can get clobbered by the C++ runtime. You need to save them. That's probably why you can't print in the slow path without crashing.

File: src/hotspot/cpu/arm/gc/shared/barrierSetAssembler_arm.cpp [hunk 1]
```diff
@@ -195,3 +198,43 @@ void BarrierSetAssembler::incr_allocated_bytes(MacroAssembler* masm, RegisterOrC
...
+
+  Label skip, guard;
+  Address thread_disarmed_addr(Rthread, in_bytes(bs_nm->thread_disarmed_offset()));
+
+  __ block_comment("nmethod_barrier begin");
+  __ ldr_label(tmp0, guard);
+
+  // Subsequent loads of oops must occur after load of guard value.
+  // BarrierSetNMethod::disarm sets guard with release semantics.
+  __ membar(MacroAssembler::LoadLoad, tmp1);
```
Comment (reinrich) [inline src/hotspot/cpu/arm/gc/shared/barrierSetAssembler_arm.cpp:221]: Is my assumption correct that ARM32 supports neither ZGC nor ShenandoahGC?
If so then no memory barrier is needed as no oops are modified concurrently with G1 (and other GCs are not concurrent).
I'd recomment to implement `BarrierSetAssembler::nmethod_patching_type()` as on AARCH64 and replace the membar with an assertion that `NMethodPatchingType::stw_instruction_and_data_patch` is in effect.
@fisk wouldn't it be good if there was a shared method to query if oops (other values?) are modified concurrently?

File: src/hotspot/cpu/arm/gc/shared/barrierSetNMethod_arm.cpp [hunk 2]
```diff
@@ -23,18 +23,119 @@
...
+}
+
+/* We're called from an nmethod when we need to deoptimize it. We do
+   this by throwing away the nmethod's frame and jumping to the
+   ic_miss stub. This looks like there has been an IC miss at the
+   entry of the nmethod, so we resolve the call, which will fall back
+   to the interpreter if the nmethod has been unloaded. */
 void BarrierSetNMethod::deoptimize(nmethod* nm, address* return_address_ptr) {
-  ShouldNotReachHere();
+
```
Comment (reinrich) [inline src/hotspot/cpu/arm/gc/shared/barrierSetNMethod_arm.cpp:87]: Again if neither ZGC nor ShenandoahGC are supported then you can leave this unimplemented. [`BarrierSetNMethod::nmethod_entry_barrier()`](https://github.com/openjdk/jdk/blob/fa0c599a003962cf3457bb49b9f771659532b5c3/src/hotspot/share/gc/shared/barrierSetNMethod.cpp#L72-L103) never returns false so the single [call site](https://github.com/openjdk/jdk/blob/fa0c599a003962cf3457bb49b9f771659532b5c3/src/hotspot/share/gc/shared/barrierSetNMethod.cpp#L181) cannot be reached unless you want to support `DeoptimizeNMethodBarriersALot` which IMO is not needed without support for ZGC/ShenandoahGC.

File: src/hotspot/cpu/arm/stubGenerator_arm.cpp [hunk 3]
```diff
@@ -2905,6 +2906,51 @@ class StubGenerator: public StubCodeGenerator {
...
+    __ pop(saved_regs);
+
+    __ cbnz(Rtemp, deoptimize_label);
+
+    __ leave();
+    __ bx(LR);
+
+    __ BIND(deoptimize_label);
+
+    __ ldr(Rtemp, Address(SP, 0));
```
Comment (reinrich) [inline src/hotspot/cpu/arm/stubGenerator_arm.cpp:2944]: If neither ZGC nor ShenandoahGC are supported then `deoptimization_label` will not be reached.

**Review by reinrich — COMMENTED** [review]
Hi Aleksei,
thanks for picking up the work on ARM32 support of nmethod entry barriers.
Is my assumption correct that ARM32 supports neither ZGC nor ShenandoahGC? If so you could simplify the implementation.
Thanks, Richard.

File: src/hotspot/cpu/arm/gc/shared/barrierSetNMethod_arm.cpp (hunk 2 above)
Comment (TheRealMDoerr) [reply src/hotspot/cpu/arm/gc/shared/barrierSetNMethod_arm.cpp:87]: Isn't there a jtreg test which triggers the deoptimization somehow?

Comment (voitylov) [reply src/hotspot/cpu/arm/gc/shared/barrierSetNMethod_arm.cpp:87]: I could not find jtreg tests that specially target deoptimization codepath but one can trigger and test it using a benchmark, e.g. dacapo lusearch with +DeoptimizeNMethodBarriersALot key.

**Comment by voitylov** [comment]
@fisk @reinrich Thank you for your suggestions. This iteration saves floating registers and can successfully log deoptimization. I'd prefer to keep the code in line with the same code for other ports and not carve out memory barriers by hand, if @reinrich would agree. This also allows a more straightforward port for additional GCs in the future, should there be any demand.

**Comment by TheRealMDoerr** [comment]
Please note that the memory barrier has caused a performance regression on PPC64 (as on AArch64): https://bugs.openjdk.org/browse/JDK-8295069
I'd probably assert `nmethod_patching_type() == NMethodPatchingType::stw_instruction_and_data_patch` instead of emitting a memory barrier because `LoadLoad` may or may not be strong enough depending on usage (e.g. generational ZGC uses instruction patching as well).

**Comment by voitylov** [comment]
The removed membar can be added later if and when support for other GCs is added to ARM32.

File: src/hotspot/cpu/arm/gc/shared/barrierSetNMethod_arm.cpp (hunk 2 above)
Comment (TheRealMDoerr) [reply src/hotspot/cpu/arm/gc/shared/barrierSetNMethod_arm.cpp:86]: ok, thanks for checking

File: src/hotspot/cpu/arm/gc/shared/barrierSetAssembler_arm.cpp (hunk 1 above)
Comment (dsamersoff) [reply src/hotspot/cpu/arm/gc/shared/barrierSetAssembler_arm.cpp]: I'm second to the proposal to move nmethod_patching_type() to shared code.

**Review by TheRealMDoerr — APPROVED** [review]
LGTM, but I'm not an arm 32 expert.

**Review by reinrich — APPROVED** [review]
The changes look good to me. I'm not familiar with arm32 details, though.
You might want to open an RFE to port JDK-8290700 to arm32.
Thanks, Richard.

**Comment by bulasevich** [comment]
Looks good!

**Comment by voitylov** [comment]
@TheRealMDoerr @reinrich @bulasevich thank you for your reviews!

**Comment by voitylov** [comment]
@fisk do you have any further comments? I'd like to have this in to continue working on Loom.

File: src/hotspot/cpu/arm/arm.ad [hunk 4]
```diff
@@ -286,6 +289,17 @@ void MachPrologNode::format( PhaseRegAlloc *ra_, outputStream *st ) const {
   if (framesize != 0) {
     st->print   ("SUB    R_SP, R_SP, " SIZE_FORMAT,framesize);
   }
+
+  if (C->stub_function() == NULL && BarrierSet::barrier_set()->barrier_set_nmethod() != NULL) {
+    st->print("ldr  t0, [guard]\n\t");
```
Comment (theRealAph) [inline src/hotspot/cpu/arm/arm.ad:294]:
```suggestion
    st->print("ldr t0, [guard]\n\t");
```

File: src/hotspot/cpu/arm/gc/shared/barrierSetAssembler_arm.cpp [hunk 5]
```diff
@@ -195,3 +198,47 @@ void BarrierSetAssembler::incr_allocated_bytes(MacroAssembler* masm, RegisterOrC
...
+
+  Register tmp0 = Rtemp;
+  Register tmp1 = R5; // wrong register here will lead to crash
+
+  if (bs_nm == NULL) {
+    return;
+  }
+
+  // The are no GCs that require memory barrier on arm32 now
+#ifndef PRODUCT
```
Comment (theRealAph) [inline src/hotspot/cpu/arm/gc/shared/barrierSetAssembler_arm.cpp:214]:
```suggestion
#ifdef ASSERT
```

File: src/hotspot/cpu/arm/gc/shared/barrierSetAssembler_arm.cpp [hunk 6]
```diff
@@ -195,3 +198,47 @@ void BarrierSetAssembler::incr_allocated_bytes(MacroAssembler* masm, RegisterOrC
   // Unborrow the Rthread
   __ sub(Rthread, Ralloc, in_bytes(JavaThread::allocated_bytes_offset()));
 }
+
+void BarrierSetAssembler::nmethod_entry_barrier(MacroAssembler* masm) {
+
+  BarrierSetNMethod* bs_nm = BarrierSet::barrier_set()->barrier_set_nmethod();
+
+  Register tmp0 = Rtemp;
+  Register tmp1 = R5; // wrong register here will lead to crash
```
Comment (theRealAph) [inline src/hotspot/cpu/arm/gc/shared/barrierSetAssembler_arm.cpp:207]: Why specifically `R5` ? Please add a comment.

File: src/hotspot/cpu/arm/arm.ad [hunk 7]
```diff
@@ -286,6 +289,17 @@ void MachPrologNode::format( PhaseRegAlloc *ra_, outputStream *st ) const {
...
+    st->print("ldr t1, [Rthread, #thread_disarmed_offset]\n\t");
+    st->print("cmp t0, t1\n\t");
+    st->print("beq skip\n\t");
+    st->print("blr #nmethod_entry_barrier_stub\n\t");
+    st->print("b skip\n\t");
+    st->print("guard: int\n\t");
+    st->print("skip:\n\t");
+  }
 }
```
Comment (theRealAph) [inline src/hotspot/cpu/arm/arm.ad:303]: I wonder if this stuff should be in `barrierSetAssembler` ? Just a thought.

File: src/hotspot/cpu/arm/gc/shared/barrierSetNMethod_arm.cpp [hunk 8]
```diff
@@ -23,18 +23,124 @@
...
+#include "runtime/frame.inline.hpp"
+#include "runtime/javaThread.hpp"
+#include "runtime/sharedRuntime.hpp"
+#include "runtime/registerMap.hpp"
+#include "utilities/align.hpp"
 #include "utilities/debug.hpp"
 
+// The constant below reflects the size of the barrier
+// in barrierSetAssembler_arm.cpp
+static const int entry_barrier_bytes = 36;
```
Comment (theRealAph) [inline src/hotspot/cpu/arm/gc/shared/barrierSetNMethod_arm.cpp:40]:
```suggestion
static const int entry_barrier_bytes = 9 * NativeInstruction::size();
```

**Review by theRealAph — APPROVED** [review]
Looks reasonable. Please consider my cleanups.

**Comment by voitylov** [comment]
Thanks @theRealAph . Your suggestions for cleanups are addressed.

**Review by TheRealMDoerr — APPROVED** [review]
Thanks for the cleanup! Still looks good. I think the PR is ready for integration. You have 3 reviews and the concerns were addressed.

**Review by fisk — APPROVED** [review]
Looks good.

**Pull request closed without merging**